File: back-end/src/main/java/pl/straburzynski/campfiresongs/service/SessionConverter.java
package pl.straburzynski.campfiresongs.service;

import com.fasterxml.jackson.core.JsonProcessingException;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import pl.straburzynski.campfiresongs.model.Session;
import pl.straburzynski.campfiresongs.model.SessionDto;

@Service
@RequiredArgsConstructor
public class SessionConverter {

    private final SongConverter songConverter;

    public SessionDto convertFromSession(Session session) {
        return SessionDto.builder()
                .id(session.getId())
                .name(session.getName())
                .password(session.getPassword())
                .song(session.isTemporary()
                        ? songConverter.parseFromSongData(session.getSongData())
                        : songConverter.convertFromSong(session.getSong()))
                .temporary(session.isTemporary())
                .songData(session.getSongData())
                .build();
    }

}

File: back-end-kotlin/src/main/kotlin/pl/straburzynski/campfiresongs/session/service/SessionConverter.kt
package pl.straburzynski.campfiresongs.session.service

import org.springframework.stereotype.Service
import pl.straburzynski.campfiresongs.session.model.Session
import pl.straburzynski.campfiresongs.session.model.SessionDto
import pl.straburzynski.campfiresongs.song.service.SongConverter


@Service
class SessionConverter(val songConverter: SongConverter) {

    fun convertFromSession(session: Session): SessionDto = SessionDto(
        session.id,
        if (session.temporary)
            session.songData?.let { songConverter.parseFromSongData(session.songData) } else
            session.song?.let { songConverter.convertFromSong(it) },
        session.name,
        session.password,
        session.temporary,
        session.songData
    )
}
